### ISSUE TITLE
Sometimes Alt gets stuck

### ISSUE THREAD

**Issue by gh658804** [stefansundin/altdrag#48]
As the title says, sometimes for unknown reasons, AltDrag thinks that Alt key (the default modifier - Left & Right Alt) is pressed while it is not. 

It occasionally happens with keyboards - Alt (or Ctrl or Shift) gets "stuck". Pressing and releasing both Alt keys together a few times makes Alt get unstuck. This is not the case with AltDrag though. It seems Alt gets stuck inside AltDrag (the program thinks the modifier is always pressed, even if it isn't). In those cases, the mouse tries to drag everything on screen, which makes working impossible. Pressing and releasing one or both Alts makes no difference. The only thing the user can do is to open the task manager using the keyboard, find AltDrag process and kill it. When re-run, everything works as it is supposed to.

**Comment by stefansundin** [comment]
This has happened to me a few times, and I haven't been able to figure out how to fix it.

One thing you can do is to press the Esc key. If AltDrag detects that the Esc key is pressed, it will stop everything it is currently doing and reset its state.

If someone can find a 100% consistent way to reproduce this bug, let me know. I need to be able to reproduce it to find a good solution.

**Comment by gh658804** [comment]
>>If someone can find a 100% consistent way to reproduce this bug, let me know. I need to be able to reproduce it to find a good solution.

Understandable. Unfortunately, i too have no idea how or why it happens.

When you say "Press Esc", i assume you mean anywhere in Windows, in any application. AltDrag does not need to have the focus, right?

**Comment by stefansundin** [comment]
> AltDrag does not need to have the focus, right?

Correct. You can try it out by dragging a window and then hitting Esc. AltDrag should stop the dragging immediately.

**Comment by gh658804** [comment]
Thanks for the info. Very useful. I'd like to mention here something i hadn't realized until now. Alt+Esc is a Windows keystroke which switches between open windows (like Alt + Tab) without showing the preview window, so if you're moving something with AltDrag and press Esc while Alt is down, the movement operation IS cancelled but the previously dragged window is also put into background. I thought it was a bug at first! :)

**Comment by stefansundin** [comment]
I decided to not cancel the Esc keypress from being sent, simply because there is a chance I might change the behavior of another program. Every time you press Esc, AltDrag resets its state, even if it's not really being used. So I simply decided not to add any logic there to block the keypress because hopefully you will not need to use it that much. :)

**Comment by dn7** [comment]
Definitely not 100% the reason, but I've often noticed it getting stuck after accidentally maximizing a window via alt + double-click.
